drivers/net/mvpp2/mrvl_ptp.c

```diff
@@ -0,0 +1,346 @@
+#define MVPP22_PTP_PACKETFORMAT(x)	(((x) & 7) << 4)
+#define MVPP22_PTP_MACTIMESTAMPINGEN	BIT(11)
+#define MVPP22_PTP_TIMESTAMPENTRYID(x)	(((x) & 31) << 12)
+#define MVPP22_PTP_TIMESTAMPQUEUESELECT	BIT(18)
+
+#define MVPP22_PTP_DESC_MASK_LOW  0xfff
+#define PTP_CLASS_V1    0x01 /* protocol version 1 */
+#define PTP_CLASS_VMASK 0x0f /* max protocol version is 15 */
+
+bool mvpp2_tx_hw_tstamp(struct mrvl_priv *priv,
+			       struct pp2_ppio_desc *tx_desc,
+			       struct rte_mbuf *mbuf)
+{
+	struct mvpp2_hwtstamp_queue *queue;
+	unsigned int mtype, type, i;
+	struct ptp_header *hdr;
+	u64 ptpdesc;
+
+#if 0
+	if (!(mbuf->ol_flags & PKT_TX_IEEE1588_TMST))
+		return false;
+
+	type = ptp_classify_raw(mbuf);
+	if (!type)
+		return false;
+
+	hdr = ptp_parse_header(mbuf, type);
+	if (!hdr)
+		return false;
+
+	skb_shinfo(mbuf)->tx_flags |= SKBTX_IN_PROGRESS;
+#else
+  type = PTP_CLASS_V1;
+#endif
+
+	ptpdesc = MVPP22_PTP_MACTIMESTAMPINGEN |
+		  MVPP22_PTP_ACTION_CAPTURE;
+	queue = &priv->tx_hwtstamp_queue[0];
+
+	switch (type & PTP_CLASS_VMASK) {
+	case PTP_CLASS_V1:
+		ptpdesc |= MVPP22_PTP_PACKETFORMAT(MVPP22_PTP_PKT_FMT_PTPV1);
+		break;
+
+#if 0
+	case PTP_CLASS_V2:
+		ptpdesc |= MVPP22_PTP_PACKETFORMAT(MVPP22_PTP_PKT_FMT_PTPV2);
+		mtype = hdr->msg_type & 15;
+		/* Direct PTP Sync messages to queue 1 */
+		if (mtype == 0) {
+			ptpdesc |= MVPP22_PTP_TIMESTAMPQUEUESELECT;
+			queue = &priv->tx_hwtstamp_queue[1];
+		}
+		break;
+#endif
+	}
+
+	/* Take a reference on the mbuf and insert into our queue */
+	i = queue->next;
+	/* Iterate over 0..31 */
+	queue->next = (i + 1) & 31;
+	rte_pktmbuf_free(queue->mbuf[i]);
+	queue->mbuf[i] = rte_pktmbuf_clone(mbuf, mbuf->pool);
+
+	ptpdesc |= MVPP22_PTP_TIMESTAMPENTRYID(i);
+
+	/*
+	 * 3:0		- PTPAction
+	 * 6:4		- PTPPacketFormat
+	 * 7		- PTP_CF_WraparoundCheckEn
+	 * 9:8		- IngressTimestampSeconds[1:0]
+	 * 10		- Reserved
+	 * 11		- MACTimestampingEn
+	 * 17:12	- PTP_TimestampQueueEntryID[5:0]
+	 * 18		- PTPTimestampQueueSelect
+	 * 19		- UDPChecksumUpdateEn
+	 * 27:20	- TimestampOffset
+	 *			PTP, NTPTransmit, OWAMP/TWAMP - L3 to PTP header
+	 *			NTPTs, Y.1731 - L3 to timestamp entry
+	 * 35:28	- UDP Checksum Offset
+	 *
+	 * stored in tx descriptor bits 75:64 (11:0) and 191:168 (35:12)
+	 */
+	tx_desc->cmds[2] &=
+		rte_cpu_to_le_32(~MVPP22_PTP_DESC_MASK_LOW);
+	tx_desc->cmds[2] |=
+		rte_cpu_to_le_32(ptpdesc & MVPP22_PTP_DESC_MASK_LOW);
+	tx_desc->cmds[4] &= rte_cpu_to_le_64(~0xffffff0000000000ULL);
+	tx_desc->cmds[4] |= rte_cpu_to_le_64((ptpdesc >> 12) << 40);
+
+	MRVL_LOG(DEBUG, "### %u-%u TX set ts desc: 0x%"PRIx64,
+			priv->ppio->pp2_id, priv->ppio->port_id, ptpdesc);
+	return true;
+}
+
+static void mvpp2_isr_handle_ptp_queue(void *arg)
+{
+	struct mrvl_priv *priv = arg;
+	struct mvpp2_hwtstamp_queue *queue;
+	struct rte_mbuf *mbuf;
+	uint8_t it;
+	uint8_t nq = 0;
+
+	queue = &priv->tx_hwtstamp_queue[nq];
+
+	while (1) {
+		u32 ts = pp2_ppio_get_outq_ts(priv->ppio, nq, &it);
```

Comment:
I could not find source for this function; in any case, it is unclear how the HW knows to associate "it" with a packet; as mvpp2_tx_hw_tstamp does not forward "i" to the HW.

Reply:
I've moved this to MUSDK as a patch because it needs access to library internals. In mvpp2_tx_hwtstamp "i" is encoded in ptp descriptor (and set to "tx_desc") and then should emerge on ptp register "r0" as "it"
see https://github.com/siklu/devices/pull/1358/files#diff-f1403f7ce1669f51f8d75231f0e513f5860db0b6e6ff02d0fd16124d024760fd